crates/api_common/src/build_response.rs

```diff
@@ -217,3 +227,225 @@ pub async fn send_local_notifs(
 
   Ok(recipient_ids)
 }
+
+#[cfg(test)]
+mod tests {
```

Comment:
You can probably remove all this, it just adds to code complication. Email and notification tests are pointless unless you can observe the recipient side.

Reply:
Strongly disagree. Not all tests need to or should be end to end. When projects grow these become ever more costly. On the other hand unit tests at the contract level keep things easy to refactor and safeguard future changes.

Reply:
Testing is good, but it shouldnt be too much. In this case it looks like some enterprise software with all the abstractions. Its too complex and will make future changes unnecessarily complicated. Rust already ensures that refactoring doesnt break things.

Reply:
Agree. Rust already ensures in-code contracts. The unit tests we *do* have, are all testing the business-logic related to the database.

We don't need to do the enterprise java thing of creating DAOs and testing every tiny detail.

Reply:
As you mentioned here, only unit tests will help with bugs in logic like e.g. using an unintended field for a check or messing up a boolean condition. The logic to map certain user fields to behavior is precisely a logical thing not at all encoded in rust types. IMHO this isn't a "tiny detail" but a major feature. In fact, I know I could have done a quick and dirty job and simply modified the core logic in 10% of time it took to write the tests. I'm not a dogmatic person, e.g. I'd rather not dismiss "enterprise bloat" for philosophical reasons. Instead, let's take a look at concrete pros and cons. Though as you mention, y'all are the maintainers and deserve the last word.

Cons:
* `send_email` is now a dynamic call, a runtime change that's trivial compared to the time it takes to actually push bytes over the wire to send the data
* Additional field in lemmy context/god object. Set and forget. Email contract hasn't changed in decades, not expected to change soon
* Tests need to change if email/notification is refactored. Possible but unlikely, the tests are only testing the contract outcomes and _not_ any intermediate private logic
Pros:
* I was initially afraid of touching this code since it had _no_ tests checking contracts. Now it does and future contributors benefit from having their changes automatically verified. This means a smaller barrier on entry to making changes. Something I expect an OS project can benefit from
* Maintainers can quickly verify the extent of testing done just looking at the code, i.e. no need to remember minutiae of manual testing. This is specially useful for niche/less common used code paths
* Faster feedback on regressions. E.g. the tests encode proof that behavior is backwards compatible, and that should remain so for future changes

As mentioned before, it really is up to y'all but as a contributor I feel much safe updating code that already has thorough automated tests. As a code maintainer it should also make reviews much simpler.

Reply:
> Additional field in lemmy context/god object. Set and forget. Email contract hasn't changed in decades, not expected to change soon

The context should stay as small as possible, because its passed around a lot, all the apub calls, all the http calls, and injected in a lot of places. Nothing but your unit tests actually need an email in the context. That should be a red flag that none of it is necessary, and adding it to the context is unecessary bloat.

And 2ndly, this doesn't make the code simpler, just needlessly more complex. The only test worth doing there, is if you were to set up some kind of internal email service where you can verify that the email were actually sent, and the receiving side matches it.

So yes, please remove all the unecessary testing and just make this the settings addition and simple logical flag for when to send.

Reply:
Bump